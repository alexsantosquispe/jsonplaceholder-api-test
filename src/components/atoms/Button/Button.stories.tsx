import type { Meta, StoryObj } from '@storybook/react';

import { Button, ButtonProps } from './Button';
import { Wrapper } from '../../../utils/stories-utils';

const meta = {
  title: 'Atoms/Button',
  component: Button,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof Button>;

export const Default: Story = (args: Partial<ButtonProps>) => {
  return (
    <div className="grid grid-cols-4 gap-4">
      <Wrapper className="flex-col">
        <Button label={args.label} />
        <span className="text-sm">Default</span>
      </Wrapper>
      <Wrapper className="flex-col">
        <Button label={args.label} isDisabled />
        <span className="text-sm">Disabled</span>
      </Wrapper>
      <Wrapper className="flex-col">
        <Button label="Secondary" isSecondary />
        <span className="text-sm">Secondary</span>
      </Wrapper>
      <Wrapper className="flex-col">
        <Button label="Secondary" isSecondary isDisabled />
        <span className="text-sm">Disabled</span>
      </Wrapper>
    </div>
  );
};

Default.args = {
  label: 'Button',
  onClick: () => alert('Button clicked!')
};
