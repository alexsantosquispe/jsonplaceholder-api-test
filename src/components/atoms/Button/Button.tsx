import cn from 'clsx';
import { ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';

export interface ButtonProps {
  label?: string;
  type?: 'button' | 'submit';
  icon?: ReactNode;
  onClick?: () => void;
  isSecondary?: boolean;
  isDisabled?: boolean;
  className?: string;
  ariaLabel?: string;
}

export const Button = ({
  label,
  type = 'button',
  onClick,
  icon = null,
  isDisabled = false,
  isSecondary = false,
  className = '',
  ariaLabel
}: ButtonProps) => {
  return (
    <button
      role="button"
      aria-disabled={isDisabled}
      aria-label={ariaLabel}
      className={twMerge(
        'flex cursor-pointer items-center justify-center gap-1 rounded-lg px-4 py-2 text-sm font-medium transition-colors duration-150 ease-in-out',
        cn({
          'bg-primary dark:text-primary text-white shadow-xs dark:bg-white/90':
            !isSecondary,
          'hover:bg-primary/80 dark:hover:bg-white/80': !isDisabled,
          'text-primary/80 hover:text-primary border border-black/10 bg-transparent shadow-xs hover:bg-gray-100 dark:border-white/15 dark:text-white dark:hover:bg-white/15':
            isSecondary,
          'text-primary/40 hover:text-primary/40 cursor-not-allowed border-black/5 hover:bg-transparent dark:border-white/10 dark:text-white/25 dark:hover:bg-transparent':
            isSecondary && isDisabled,
          'bg-primary/65 cursor-not-allowed dark:bg-white/25':
            !isSecondary && isDisabled
        }),
        className
      )}
      disabled={isDisabled}
      type={type}
      onClick={onClick}
    >
      {icon && <span className="size-4">{icon}</span>}
      {label && label}
    </button>
  );
};
