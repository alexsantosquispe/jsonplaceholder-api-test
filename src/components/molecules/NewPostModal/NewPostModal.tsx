import { NewPostForm } from './components/NewPostForm/NewPostForm';
import { useCreateNewPost } from './hooks/useCreateNewPost';
import { Modal } from '../../atoms/Modal/Modal';
import { ModalProps } from '../../atoms/Modal/Modal.types';

export const NewPostModal = ({ onClose }: Pick<ModalProps, 'onClose'>) => {
  const { createPostTrigger, isLoading } = useCreateNewPost(onClose);

  return (
    <Modal
      onClose={onClose}
      title="New Post"
      acceptLabel="Create Post"
      className={{ container: 'md:w-[32rem]', body: 'pb-4 md:pb-2' }}
    >
      <NewPostForm isLoading={isLoading} onCreatePost={createPostTrigger} />
    </Modal>
  );
};
