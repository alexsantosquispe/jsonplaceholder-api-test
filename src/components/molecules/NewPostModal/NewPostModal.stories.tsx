import type { Meta, StoryObj } from '@storybook/react';
import { QueryClientProvider } from '@tanstack/react-query';
import { useState } from 'react';

import NewPostModal from './NewPostModal';
import { AlertProvider } from '../../../contexts/AlertProvider';
import queryClient from '../../../services/queryClient';
import { Wrapper } from '../../../utils/stories-utils';
import { AlertStack } from '../../atoms/AlertStack/AlertStack';
import { Button } from '../../atoms/Button/Button';

const NewPostModalWrapper = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <AlertProvider>
      <QueryClientProvider client={queryClient}>
        <Wrapper className="flex-col items-start justify-start">
          <Button label="Open New Post Modal" onClick={() => setIsOpen(true)} />
          <div className="flex items-baseline pt-6 text-sm text-neutral-600">
            <strong>Note:&nbsp;&nbsp;</strong>Press the
            <pre className="text-primary mx-1 rounded bg-gray-100 px-2 py-0.5">
              Esc
            </pre>
            key in order to close the modaL
          </div>
          {isOpen && <NewPostModal onClose={() => setIsOpen(false)} />}
        </Wrapper>
        <AlertStack />
      </QueryClientProvider>
    </AlertProvider>
  );
};

const meta = {
  title: 'Molecules/NewPostModal',
  component: NewPostModal
} satisfies Meta<typeof NewPostModal>;

export default meta;

type Story = StoryObj<typeof NewPostModal>;

export const Default: Story = () => {
  return <NewPostModalWrapper />;
};

Default.args = {};
