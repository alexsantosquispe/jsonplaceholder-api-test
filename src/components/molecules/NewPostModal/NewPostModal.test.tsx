import { fireEvent, render, screen } from '@testing-library/react';
import { useState } from 'react';

import NewPostModal from './NewPostModal';
import {
  AlertProviderWrapper,
  QueryClientProviderWrapper
} from '../../../utils/test-utils';
import { Button } from '../../atoms/Button/Button';

const NewPostModalWrapper = () => {
  const [isOpen, setIsOpen] = useState(false);
  return (
    <AlertProviderWrapper>
      <QueryClientProviderWrapper>
        <>
          <Button label="Open modal" onClick={() => setIsOpen(true)} />
          {isOpen && <NewPostModal onClose={() => setIsOpen(false)} />}
        </>
      </QueryClientProviderWrapper>
    </AlertProviderWrapper>
  );
};

describe('NewPostModal', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the component correctly', () => {
    const component = render(<NewPostModalWrapper />);

    expect(component).toMatchSnapshot();
  });

  describe('Behavior', () => {
    it('it should open/close the NewPostModal', () => {
      render(<NewPostModalWrapper />);

      const openModalButton = screen.getByRole('button');

      expect(openModalButton).toBeInTheDocument();

      fireEvent.click(openModalButton);

      expect(screen.getByRole('form')).toBeInTheDocument();
    });
  });
});
