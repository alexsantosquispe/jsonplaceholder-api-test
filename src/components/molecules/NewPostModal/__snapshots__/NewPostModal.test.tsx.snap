// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPostModal should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="fixed top-[8.5rem] z-40 flex w-full flex-col items-center gap-y-2 md:top-[4.5rem]"
      />
      <button
        aria-disabled="false"
        class="flex cursor-pointer items-center justify-center gap-1 rounded-lg px-4 py-2 text-sm font-medium transition-colors duration-150 ease-in-out bg-primary dark:text-primary text-white shadow-xs dark:bg-white/90 hover:bg-primary/80 dark:hover:bg-white/80"
        role="button"
        type="button"
      >
        Open modal
      </button>
    </div>
  </body>,
  "container": <div>
    <div
      class="fixed top-[8.5rem] z-40 flex w-full flex-col items-center gap-y-2 md:top-[4.5rem]"
    />
    <button
      aria-disabled="false"
      class="flex cursor-pointer items-center justify-center gap-1 rounded-lg px-4 py-2 text-sm font-medium transition-colors duration-150 ease-in-out bg-primary dark:text-primary text-white shadow-xs dark:bg-white/90 hover:bg-primary/80 dark:hover:bg-white/80"
      role="button"
      type="button"
    >
      Open modal
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
